core/src/main/java/com/graphhopper/routing/InstructionsFromEdges.java

```diff
@@ -390,7 +394,7 @@ private int getTurn(EdgeIteratorState edge, int baseNode, int prevNode, int adjN
             forceInstruction = true;
         }
 
-        InstructionsOutgoingEdges outgoingEdges = new InstructionsOutgoingEdges(prevEdge, edge, encoder, crossingExplorer, nodeAccess, prevNode, baseNode, adjNode);
+        InstructionsOutgoingEdges outgoingEdges = new InstructionsOutgoingEdges(prevEdge, edge, encoder, maxSpeedEnc, crossingExplorer, nodeAccess, prevNode, baseNode, adjNode);
```

Comment:
It is not really related to this issue, but maybe you can have a look if we can solve it here:

the crossingExplorer might have a problem. It is created like:
```
crossingExplorer = graph.createEdgeExplorer(DefaultEdgeFilter.allEdges(encoder));
```

But in the InstructionsOutgoingEdges constructor we do:

allOutgoingEdges.add(tmpEdge);

Couldn't it be that some of these edges are also incoming edges and so at least the name allOutgoingEdges is not really correct?

Reply:
But maybe this should go into another issue ...

Reply:
The whole code looks like:

```java
            if (edgeIter.getAdjNode() != prevNode && edgeIter.getAdjNode() != adjNode) {
                tmpEdge = edgeIter.detach(false);
                allOutgoingEdges.add(tmpEdge);
                if (tmpEdge.get(accessEnc)) {
                    allowedOutgoingEdges.add(tmpEdge);
                }
            }
```

Shouldn't the if filter out the edges that are on the path, so it's only the outgoing edges, that are not on the path?

Yes, the naming could definitely be improved, as it's not strictly clear what edges are in this collection.

What name should we choose?
allUnusedOutgoingEdges
allOutgoingEdgesThatAreNotOnThePath (a bit too long :D)

It's not, possible turns, as it also contains roads you are not allowed to turn on. Possible turns are stored in `allowedOutgoingEdges`.

Reply:
What I meant was more: could it be that `allOutgoingEdges` also includes the incoming edges (that are not on the path)?

As the explorer is created via `createEdgeExplorer(DefaultEdgeFilter.allEdges(encoder));` and not via `DefaultEdgeFilter.outEdges`

Reply:
Yes, IIRC it was designed like this. For this we have the allowedOutgoingEdges, these should only contain the edges that we could legally turn onto (ignoring turn restrictions - I think).

We use this for example [here](https://github.com/graphhopper/graphhopper/blob/58cdbdfcfd9862b354e9066c0ff0b67ab3179888/core/src/main/java/com/graphhopper/routing/InstructionsFromEdges.java#L398). Even though, we are only allowed to turn on one alternative, there is another edge that could potentially be mistaken for a turn, so we create a turn instruction to be sure that this is a turn.

Reply:
Also see the [todo](https://github.com/graphhopper/graphhopper/blob/58cdbdfcfd9862b354e9066c0ff0b67ab3179888/core/src/main/java/com/graphhopper/routing/InstructionsFromEdges.java#L402) a bit further down :). Maybe we should do the oneway check additionally?

Reply:
Probably just the name of `allOutgoingEdges` is what confuses me. Should we rename it into `edgeList` and the variable `allowedOutgoingEdges` into `outgoingEdgeList`?

Reply:
It's not the edgeList either. It's the edgeList of a junction without the streets that are used by the route. Maybe remainingEdgeList?

Reply:
This is a bit confusing ... or maybe offPathEdges and outgoingOffPathEdges? Not sure. But this also does not reflect the problem that only vehicle-specific off-the-path edges are included...

Reply:
Let's solve this is a different issue.

Reply:
> But this also does not reflect the problem that only vehicle-specific off-the-path edges are included...

Yes indeed, should we create a new issue for this?

Reply:
Or should I just rename it to a neutral filteredEdges and filteredOutgoingEdges?

Reply:
Personally, I am a fan of speaking names :). So I'd prefer a name that explains the meaning of these edges. So I would ask myself what kind of edges these are and would read the comments :). But if you think, the current name leads to potential misunderstandings of the code, we should rename it and then better give it a more generic name than a name that could be misunderstood :).

We should still look into the issues of non vehicle accessible edges to be considered for the instruction generation. And maybe also oneways, turn restrictions, etc.

But maybe we do this, when we find a good test scenario where the instructions are not correct?

Reply:
> So I'd prefer a name that explains the meaning of these edges.

But sometimes it is not really possible to put all the meaning in it ... i.e. it will get pretty nasty here like outgoingVehicleSpecificOffPathEdges :D and keeping the name is confusing - at least for me :)

> We should still look into the issues of non vehicle accessible edges to be considered for the instruction generation. 

Ok, yes. We can create such an issue. For car it shouldn't be a problem as ignoring bike-only paths is intended, but for bike ignoring car-only roads might be confusing, but I don't know.

Reply:
> and keeping the name is confusing - at least for me :)

Yes, in this case, let's rename it :)

> but for bike ignoring car-only roads might be confusing, but I don't know.

Yes indeed, or maybe for some special vehicles like scooters, wheelchairs, trucks, etc.